policy_agent/policy_agent.py

```diff
@@ -0,0 +1,354 @@
+        return "%s/%s" % (self.namespace, self.name)
+
+    def sync(self):
+        ns_tag = "namespace_%s" % self.namespace
+
+        if _datastore_client.profile_exists(ns_tag) and self.ep_id:
+            _datastore_client.set_profiles_on_endpoint(
+                [ns_tag], endpoint_id=self.ep_id)
+            return True
+        else:
+            _log.error("Pod Resource %s found before Namespace Resource %s" % (
+                self.name, self.namespace))
+            return False
+
+
+class Service(Resource):
+
+    def from_json(self, json):
+        self.kind = "Service"
+        self.uid = json["metadata"]["uid"]
+        self.name = json["metadata"]["name"]
+        self.namespace = json["metadata"]["namespace"]
+
+
+class Endpoints(Resource):
+
+    def from_json(self, json):
+        self.kind = "Endpoints"
+        self.selfLink = json["metadata"]["selfLink"]
+        self.items = json["items"]
+
+    def get_key(self):
+        return self.selfLink
+
+
+class Namespace(Resource):
+
+    def from_json(self, json):
+        self.kind = "Namespace"
+        try:
+            self.uid = json["metadata"]["uid"]
+            self.name = json["metadata"]["name"]
+        except KeyError:
+            _log.error("Namespace does not have uid or name")
+
+        try:
+            self.policy = json["metadata"]["annotations"][POLICY_ANNOTATION_KEY]
+        except KeyError:
+            _log.warning("Namespace does not have policy, assumed closed")
+            self.policy = "closed"
+
+    def sync(self):
+        # Derive NS tag
+        ns_tag = "namespace_%s" % self.name
+
+        # Add Tags
+        profile_path = PROFILE_PATH % {"profile_id": ns_tag}
+        _datastore_client.etcd_client.write(profile_path + "tags", '["%s"]' % ns_tag)
+
+        # Determine rule set based on policy
+        default_allow = Rule(action="allow")
+
+        if self.policy == "open":
+            rules = Rules(id=ns_tag,
+                      inbound_rules=[default_allow],
+                      outbound_rules=[default_allow])
+            _log.info("Applying Open Rules to NS Profile %s" % ns_tag)
+
+        elif self.policy == "closed":
+            rules = Rules(id=ns_tag,
+                      inbound_rules=[Rule(action="allow", src_tag=ns_tag)],
+                      outbound_rules=[default_allow])
+            _log.info("Applying Closed Rules to NS Profile %s" % ns_tag)
+
+        else:
+            _log.error(
+                "Namespace %s policy is neither open nor closed" % self.name)
+            sys.exit(1)
+
+        # Write rules to profile
+        _datastore_client.etcd_client.write(profile_path + "rules", rules.to_json())
+        return True
+
+
+def _keep_watch(queue, path):
+    """
+    Called by watcher threads. Adds watch events to Queue
+    """
+    response = _get_api_stream("watch/%s" % path)
+    for line in response.iter_lines():
+        if line:
+            queue.put(line)
+
+
+def _get_api_token():
```

Comment:
Fine to copy this over for now, but we don't want to duplicate code like this in the long run - this should be common between the plugin and the agent.